blockchain/chain_sync/src/sync.rs

```diff
@@ -0,0 +1,493 @@
+struct MsgMetaData {
+    balance: BigUint,
+    sequence: u64,
+}
+
+impl<'db, DB> ChainSyncer<'db, DB, HamtStateTree>
+where
+    DB: BlockStore,
+    // BS: BlockSyncProvider,
+{
+    pub fn new(
+        db: &'db DB,
+        network_send: Sender<NetworkMessage>,
+        network_rx: Receiver<NetworkEvent>,
+    ) -> Result<Self, Error> {
+        // TODO import genesis from storage
+        let _genesis = Tipset::new(vec![BlockHeader::default()])?;
+
+        // TODO change from being default when impl
+        let sync_manager = SyncManager::default();
+
+        let chain_store = ChainStore::new(db);
+
+        let state_manager = StateManager::new(db, HamtStateTree::default());
+
+        let network = SyncNetworkContext::new(network_send);
+
+        Ok(Self {
+            _state: SyncState::Stalled,
+            state_manager,
+            chain_store,
+            network,
+            _genesis,
+            sync_manager,
+            network_rx,
+            bad_blocks: LruCache::new(1 << 15),
+        })
+    }
+
+    /// Starts syncing process
+    pub async fn sync(&mut self) -> Result<(), Error> {
+        let mut nw = self.network_rx.clone().fuse();
+        loop {
+            select! {
+                network_msg = nw.next().fuse() => match network_msg {
+                    // TODO remove this comment (testing use)
+                    // Some(NetworkEvent::PubsubMessage{source, ..}) => {
+                    //     let req_id = self.network.blocksync_headers("QmUbiXETSptmbSYNtvCWEBYqXDKGwiBCCFWgJn1ebGiFiV".parse().unwrap(), self._genesis.key(), 10).await;
+                    // },
+                    Some(event) => println!("received some other event: {:?}", event),
+                    None => break,
+                }
+            }
+        }
+
+        info!("Starting chain sync");
+
+        // Get heaviest tipset from storage to sync toward
+        let heaviest = self.chain_store.heaviest_tipset();
+
+        // TODO remove this and retrieve head from storage
+        let head = Tipset::new(vec![BlockHeader::default()]).unwrap();
+
+        // Sync headers from network from head to heaviest from storage
+        let headers = self.sync_headers_reverse(head, &heaviest).await?;
+
+        // Persist header chain pulled from network
+        self.persist_headers(&headers)?;
+
+        Ok(())
+    }
+
+    /// informs the syncer about a new potential tipset
+    /// This should be called when connecting to new peers, and additionally
+    /// when receiving new blocks from the network
+    pub fn inform_new_head(&self, from: PeerId, fts: FullTipset) -> Result<(), Error> {
+        // check if full block is nil and if so return error
+        if fts.blocks().is_empty() {
+            return Err(Error::NoBlocks);
+        }
+        // validate message data
+        for block in fts.blocks() {
+            self.validate_msg_data(block)?;
+        }
+        // TODO send pubsub message indicating incoming blocks
```

Comment:
I might be mistaken but since `network_rx` was added in this PR could we complete this TODO by informing the network of the incoming block?

Reply:
hmm yes good idea, I had assumed it would happen outside of this, but the point of it happening here is because it has to be validated first? Seems redundant to repropogate all blocks in a validated tipset when those would have already been propogated per block

Reply:
Yeah, I see that now.